subsys/usb/device_next/class/usbd_midi.c

```diff
@@ -0,0 +1,786 @@
+					(MIDI_1_0_UP_TO_128_BITS),                                \
+					(USE_MIDI_CI)                                             \
+				))                                                                \
+			))                                                                        \
+		)
+
+/* Group Terminal Block unique identification number with a trailing comma
+ * if that block is bidirectional or of given terminal type; otherwise empty
+ */
+#define GRPTRM_BLOCK_ID_SEP_IF(node, ttype)                                    \
+	IF_ENABLED(                                                            \
+		UTIL_OR(DT_ENUM_HAS_VALUE(node, terminal_type, bidirectional), \
+			DT_ENUM_HAS_VALUE(node, terminal_type, ttype)),        \
+		(GRPTRM_BLOCK_ID(node),))
+
+/* All unique identification numbers of output+bidir group terminal blocks */
+#define GRPTRM_OUTPUT_BLOCK_IDS(inst) \
+	DT_INST_FOREACH_CHILD_VARGS(inst, GRPTRM_BLOCK_ID_SEP_IF, output_only)
+
+/* All unique identification numbers of input+bidir group terminal blocks */
+#define GRPTRM_INPUT_BLOCK_IDS(inst) \
+	DT_INST_FOREACH_CHILD_VARGS(inst, GRPTRM_BLOCK_ID_SEP_IF, input_only)
+
+#define N_INPUTS(inst)  sizeof((uint8_t[]){GRPTRM_INPUT_BLOCK_IDS(inst)})
+#define N_OUTPUTS(inst) sizeof((uint8_t[]){GRPTRM_OUTPUT_BLOCK_IDS(inst)})
+
+#define USBD_MIDI_VALIDATE_GRPTRM_BLOCK(node)                              \
+	BUILD_ASSERT(DT_REG_ADDR(node) < 16,                               \
+		     "Group Terminal Block address must be within 0..15"); \
+	BUILD_ASSERT(DT_REG_ADDR(node) + DT_REG_SIZE(node) <= 16,          \
+		     "Too many Group Terminals in this Block");
+
+#define USBD_MIDI_VALIDATE_INSTANCE(inst) \
+	DT_INST_FOREACH_CHILD(inst, USBD_MIDI_VALIDATE_GRPTRM_BLOCK)
+
+#define USBD_MIDI2_INIT_GRPTRM_BLOCK_DESCRIPTOR(node)                       \
+	{                                                                   \
+		.bLength = sizeof(struct usb_midi_grptrm_block_descriptor), \
+		.bDescriptorType = CS_GR_TRM_BLOCK,                         \
+		.bDescriptorSubtype = GR_TRM_BLOCK,                         \
+		.bGrpTrmBlkID = GRPTRM_BLOCK_ID(node),                      \
+		.bGrpTrmBlkType = GRPTRM_BLOCK_TYPE(node),                  \
+		.nGroupTrm = DT_REG_ADDR(node),                             \
+		.nNumGroupTrm = DT_REG_SIZE(node),                          \
+		.iBlockItem = 0,                                            \
+		.bMIDIProtocol = GRPTRM_PROTOCOL(node),                     \
+		.wMaxInputBandwidth = 0x0000,                               \
+		.wMaxOutputBandwidth = 0x0000,                              \
+	}
+
+#define USBD_MIDI2_GRPTRM_TOTAL_LEN(inst)                 \
+	sizeof(struct usb_midi_grptrm_header_descriptor)  \
+	+ DT_INST_CHILD_NUM_STATUS_OKAY(inst)             \
+	* sizeof(struct usb_midi_grptrm_block_descriptor)
+
+#define USBD_MIDI_DEFINE_DESCRIPTORS(inst)                                               \
```

Comment:
I prefer just `n` and want it to be consistent with other class implementations `#define USBD_MIDI_DEFINE_DESCRIPTORS(n)`